src/blocknative.rs

```diff
@@ -0,0 +1,297 @@
+use super::{linear_interpolation, GasPriceEstimating, Transport};
+use crate::CachedResponse;
+use anyhow::{anyhow, Context, Result};
+use serde::Deserialize;
+use std::{
+    convert::TryInto,
+    sync::{Arc, Mutex},
+    time::{Duration, Instant},
+};
+use tokio::task::{self, JoinHandle};
+
+// Gas price estimation with https://www.blocknative.com/gas-estimator , api https://docs.blocknative.com/gas-platform#example-request .
+
+const API_URI: &str = "https://api.blocknative.com/gasprices/blockprices";
+
+const TIME_PER_BLOCK: Duration = Duration::from_secs(15);
+const RATE_LIMIT: Duration = Duration::from_secs(10);
+const CACHED_RESPONSE_VALIDITY: Duration = Duration::from_secs(60);
+
+#[derive(Debug, Deserialize, Clone, Default)]
+#[serde(rename_all = "camelCase")]
+struct EstimatedPrice {
+    confidence: f64,
+    price: f64,
+    max_priority_fee_per_gas: f64,
+    max_fee_per_gas: f64,
+}
+
+#[derive(Debug, Deserialize, Clone, Default)]
+#[serde(rename_all = "camelCase")]
+struct BlockPrice {
+    estimated_prices: Vec<EstimatedPrice>,
+}
+
+#[derive(Debug, serde::Deserialize, Clone, Default)]
+#[serde(rename_all = "camelCase")]
+struct Response {
+    block_prices: Vec<BlockPrice>,
+}
+
+struct Request<T> {
+    transport: T,
+    header: http::header::HeaderMap,
+}
+
+impl<T: Transport> Request<T> {
+    async fn gas_price(&self) -> Result<Response> {
+        self.transport
+            .get_json(API_URI, self.header.clone())
+            .await
+            .context("failed to get blocknative gas price")
+    }
+}
+
+pub struct BlockNative {
+    cached_response: Arc<Mutex<CachedResponse<Response>>>,
+    handle: JoinHandle<()>,
+}
+
+impl Drop for BlockNative {
+    fn drop(&mut self) {
+        self.handle.abort();
+        self.cached_response = Default::default();
+    }
+}
+
+impl BlockNative {
+    pub async fn new<T: Transport + 'static>(
+        transport: T,
+        header: http::header::HeaderMap,
+    ) -> Result<Self> {
+        let cached_response: Arc<Mutex<CachedResponse<Response>>> = Default::default();
+        let cached_response_clone = cached_response.clone();
+
+        //send one request to initially populate the cached response
+        let request = Request { transport, header };
+        match request.gas_price().await {
+            Ok(response) => match cached_response_clone.lock() {
```

Comment:
No need to match on being unable to lock the mutex. This can only fail if we have previously panicked while holding a lock which is not a situation we need to handle. So just unwrap. Same for the task below and when getting the estimate.

Reply:
Yes, fixed.